### PR TITLE
Async Code V0

### PR DESCRIPTION
**Pull request by davinchia**
## What
Split out the smallest set of reasonable changes from #26086 .

My goal was to split out the interface, as well as show how the interface it's meant to be used.

Follow up PRs:
* Split out classes from `BufferManager` and add more tests there.
* Add in the AsyncConsumer with tests.
* Add in the StagingConsumer factory.

## How
Split out the classes and add javadocs.

From original branch:
* renamed `StreamDestinationFlusher` to `DestinationFlushFunction`.
* renamed `UploadWorkers` to `FlushWorkers`.
* also split out the MemoryManager class to its own class.

Did not touch any tests.

Some formatting changes.

## Recommended reading order
* DestinationFlushFunction.java and FlushWorkers.java for the javadocs.
* Everything else is as is.

## 🚨 User Impact 🚨
No user impact since all of this is in a new module.

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+
+@Slf4j
+public class BufferManager {
```
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java:25]: this class can be further split and tested. will do so in a follow up PR

**Comment by davinchia** [comment]
PTAL @ryankfu @cgardens, this is the first set of PRs to get our code from #26086 into master.

With this first PR, I want to get the smallest reasonable set into master, with a focus on documenting the interfaces. 

I was hoping we could do this more piecemeal, however there is really no way since all these classes are somewhat related.

Charles, I know you were hoping to comment and rewrite the queue bits. Feel free to do so here or in a follow up PR. My preference is a follow up PR to keep things small. The queue bits are well tested so I don't believe having this in master is 'that bad'.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java:1]: I am happy to take naming suggestions.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DestinationFlushFunction.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DestinationFlushFunction.java:1]: I am happy to take naming suggestions.

**Review by ryankfu — APPROVED** [review]
I'm happy with this V0 merge, since it doesn't include any of the `StagingConsumerFactory` changes it doesn't affect snowflake directly. Agree that comments and additional breaking out the classes can be in follow up PR

It does feel a little weird that there's Postgres/MySQL changes in here but those look like `spotlessJava` linting

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java
Comment (ryankfu) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java:1]: Maybe `AsyncDataUploader`? As suggested by ChatGPT 😛

**Review by cgardens — APPROVED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java
Comment (davinchia) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java:1]: Decent idea. My thought was to keep this in line with the interface name of `DestinationFlushFunction`. Let's iterate on this next week.

**Comment by davinchia** [comment]
I am merging this in as it doesn't affect any actual connectors.

**Pull request merged**